watchtower-plugin/src/dbm.rs

```diff
@@ -216,29 +216,39 @@ impl DBM {
         &self,
         tower_id: TowerId,
         user_id: UserId,
-    ) -> Option<RegistrationReceipt> {
-        let mut stmt = self
-            .connection
-            .prepare(
-                "SELECT available_slots, subscription_start, subscription_expiry, signature
-                    FROM registration_receipts 
-                    WHERE tower_id = ?1 AND subscription_expiry = (SELECT MAX(subscription_expiry) 
-                        FROM registration_receipts 
-                        WHERE tower_id = ?1)",
-            )
-            .unwrap();
+        subscription_start: Option<u32>,
+        subscription_expiry: Option<u32>,
+    ) -> Vec<RegistrationReceipt> {
+        let mut query = "SELECT available_slots, subscription_start, subscription_expiry, signature FROM registration_receipts WHERE tower_id = ?1".to_string();
 
-        stmt.query_row([tower_id.to_vec()], |row| {
-            let slots: u32 = row.get(0).unwrap();
-            let start: u32 = row.get(1).unwrap();
-            let expiry: u32 = row.get(2).unwrap();
-            let signature: String = row.get(3).unwrap();
+        let tower_id_encoded = tower_id.to_vec();
+        let mut params: Vec<&dyn ToSql> = vec![&tower_id_encoded];
 
-            Ok(RegistrationReceipt::with_signature(
-                user_id, slots, start, expiry, signature,
-            ))
-        })
-        .ok()
+        if subscription_expiry.is_none() {
+            query.push_str(" AND subscription_expiry = (SELECT MAX(subscription_expiry) FROM registration_receipts WHERE tower_id = ?1)")
+        } else {
+            query.push_str(" AND subscription_start>=?2 AND subscription_expiry <=?3");
+            params.push(&subscription_start);
+            params.push(&subscription_expiry)
+        }
+        let mut stmt = self.connection.prepare(&query).unwrap();
+
+        let receipts = stmt
+            .query_map(params.as_slice(), |row| {
+                let slots: u32 = row.get(0)?;
+                let start: u32 = row.get(1)?;
+                let expiry: u32 = row.get(2)?;
+                let signature: String = row.get(3)?;
+
+                Ok(RegistrationReceipt::with_signature(
+                    user_id, slots, start, expiry, signature,
+                ))
+            })
+            .unwrap()
+            .collect::<Result<Vec<_>, _>>()
```

Comment:
I think we can simplify this to `Result<_, _>`. I'm not really sure why the annotation is needed tbh.

Any ideas @mariocynicys?

Reply:
I would have written it like this:
```rust
let receipts = stmt
    .query_map(params.as_slice(), |row| {
        let slots: u32 = row.get(0)?;
        let start: u32 = row.get(1)?;
        let expiry: u32 = row.get(2)?;
        let signature: String = row.get(3)?;

        Ok(RegistrationReceipt::with_signature(
            user_id, slots, start, expiry, signature,
        ))
    })
    .unwrap() // MappedRows<|&Row| -> Result<RegistrationReceipt, E>>
    .map(|r| r.unwrap())
    .collect();
```

It's pretty interesting though how the type after the `.unwrap()` (`MappedRows<|&Row| -> Result<RegistrationReceipt, E>>` which is equivalent to a `Vec<Result<RegistrationReceipt, E>>` when collected) could be collected into `Result<Vec<RegistrationReceipt>, E>`.

Reply:
> > IIRC if two subscriptions overlapped, the user was only supposed to be storing the latest receipt. However, I just tested this and two receipts were returned where receipt_a ranged Ta-Tb, and receipt_b ranged Ta-Tc with c>b.
> 
> Revising this, we made the merger on the tower side and not the client side. The client side though gets it's subscription summary in memory updated to the merger of the two overlapping receipts, but all the receipts are stored in the DB nonetheless. We should delete those receipts since the last one already covers all of the past overlapping ones, so to get a cleaner output from `getregistrationreceipt`.

Agreed, I'll open an issue to cover this